kernel/frame_allocator/src/lib.rs

```diff
@@ -861,56 +1038,66 @@ fn contains_any(
     false
 }
 
-
-/// Adds the given `frames` to the given `list` as a Chunk of reserved frames. 
+/// Adds the given `frames` to the given `regions_list` and `frames_list` as a chunk of reserved frames. 
 /// 
-/// Returns the range of **new** frames that were added to the list, 
+/// Returns the range of **new** frames that were added to the lists, 
 /// which will be a subset of the given input `frames`.
 ///
 /// Currently, this function adds no new frames at all if any frames within the given `frames` list
 /// overlap any existing regions at all. 
 /// TODO: handle partially-overlapping regions by extending existing regions on either end.
-fn add_reserved_region(
-    list: &mut StaticArrayRBTree<Chunk>,
+fn add_reserved_region_to_lists(
+    regions_list: &mut StaticArrayRBTree<PhysicalMemoryRegion>,
+    frames_list: &mut StaticArrayRBTree<FreeFrames>,
     frames: FrameRange,
 ) -> Result<FrameRange, &'static str> {
 
-    // Check whether the reserved region overlaps any existing regions.
-    match &mut list.0 {
-        Inner::Array(ref mut arr) => {
-            for chunk in arr.iter().flatten() {
-                if let Some(_overlap) = chunk.overlap(&frames) {
-                    // trace!("Failed to add reserved region {:?} due to overlap {:?} with existing chunk {:?}",
-                    //     frames, _overlap, chunk
-                    // );
-                    return Err("Failed to add reserved region that overlapped with existing reserved regions (array).");
+    // first check the regions list for overlaps and proceed only if there are none.
+    if !contains_any(&regions_list, &frames){
```

Comment:
it'd be nicer to read and produce less of a large diff if you did an early return pattern, e.g.,
```rust
if contains_any(...) {
    return Err(...);
}
```